app/controllers/OAuthDataHandler.scala

```diff
@@ -0,0 +1,174 @@
+package controllers
+
+import java.util.Date
+import javax.inject.Inject
+
+import business.models._
+import business.services._
+import business.repositories._
+import play.api.libs.json.{JsValue, JsObject}
+import play.api.mvc.Request
+
+import scala.collection.mutable.ArrayBuffer
+import scala.concurrent.ExecutionContext.Implicits._
+import scala.concurrent.Future
+import scalaoauth2.provider._
+
+trait MyOAuth2Provider extends OAuth2Provider {
+  def getMyParam[A](request: Request[A]): Map[String, Seq[String]] = {
+    request.body match {
+      case body: play.api.mvc.AnyContentAsJson if body.asJson.isDefined => {
+        body.asJson.get.as[JsObject].fields.toMap[String, JsValue].map{
+          case (k, v) => k -> ArrayBuffer(v.as[String])
+        }
+      }
+      case _ => getParam(request)
+    }
+  }
+
+  override implicit def play2oauthRequest[A](request: Request[A]): AuthorizationRequest = {
+    val param: Map[String, Seq[String]] = getMyParam(request)
+    AuthorizationRequest(request.headers.toMap, param)
+  }
+}
+
+trait IOAuthDataHandlerFactory {
+  def getInstance(): OAuthDataHandler
+}
+
+class OAuthDataHandlerFactory @Inject() (

```

Comment:
Do you use like frameworkless DI pattern (combined with MacWire for eg) ? I've seen a presentation about it.